### PR TITLE
vpr: added optional disable errors and suppress warnings

### PR DESCRIPTION
**Pull request by acomodi**
Signed-off-by: Alessandro Comodi <acomodi@antmicro.com>

#### Description

This PR adds the functionality to do the following:
- Disable errors by demoting them to warnings;
- Suppress noisy warnings and redirect them to a custom file.

To enable this feature there are two new command line options:
- Warning suppression:
```
--suppress_warnings <output_filename>,<func1>:<func2>:<func3>
```
If the `<output_filename>` is not provided the warnings will be output on stdout as usual.

- Error disabling:
```
--disable_errors <func1>:<func2>:<func3>
```

`<funcN>` is the name of the function for which the warning/error will be suppressed/disabled, therefore, if a function name is specified in the list, all its warnings/errors are going to be disabled.

#### Motivation and Context

There are two different reasons why this PR is needed:

1. There are some errors (`VPR_THROWs`) which prevent development from going further and, instead of changing the `VPR_THROW` into a `VTR_LOG_WARN`, it would be very handy to have a mechanism as the one proposed in this PR;
2. It often happens that some warnings are repeated thousands of times and are related to the same issue. This results in having a hard time reading the output logs to find possible bugs.

A note is that these command line options should be used only if the reason behind the disabling/suppression of an error/warning is clear to the developer.

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by acomodi** [comment]
@kmurray @vaughnbetz FYI

File: libs/libvtrutil/src/vtr_log.cpp [hunk 1]
```diff
@@ -14,3 +18,33 @@ void set_log_file(const char* filename) {
 }
 
 } // namespace vtr
+
+void add_warnings_to_suppress(std::string function_name) {
+    warnings_to_suppress.insert(function_name);
+}
+
+void set_noisy_warn_log_file(const char* log_file_name) {
```
Comment (vaughnbetz) [inline libs/libvtrutil/src/vtr_log.cpp]: Should comment what you're doing in this function (making sure the log file is empty).

File: libs/libvtrutil/src/vtr_log.cpp [hunk 2]
```diff
@@ -14,3 +18,33 @@ void set_log_file(const char* filename) {
...
+    std::ofstream log;
+    log.open(log_file_name, std::ifstream::out | std::ifstream::trunc);
+    log.close();
+    noisy_warn_log_file = std::string(log_file_name);
+}
+
+void suppress_warning(const char* pszFileName, unsigned int lineNum, const char* pszFuncName, const char* pszMessage, ...) {
```
Comment (vaughnbetz) [inline libs/libvtrutil/src/vtr_log.cpp]: I think this should be called print_or_suppress_warning or some such to be more clear (it's responsible for printing the warning if it is not suppressed). Should also have a comment saying what it does, either here or at its declaration.

File: libs/libvtrutil/src/vtr_log.cpp (hunk 1 above)
Comment (acomodi) [reply libs/libvtrutil/src/vtr_log.cpp]: Done. I've added the comment above the function declaration

File: libs/libvtrutil/src/vtr_log.cpp (hunk 2 above)
Comment (acomodi) [reply libs/libvtrutil/src/vtr_log.cpp]: Done.

**Comment by acomodi** [comment]
@vaughnbetz Ping

**Pull request merged**